### PR TITLE
`store` and `load` docs

### LINKED ISSUE #346: `store`, `load` cheatcodes
Cheatcodes manipulating storage

**Need resolving before this is started:**

- [x] #512 

## Important sidenote:

We should also, consider making a PR to the compiler to avoid having to use 
`create_state_for_testing()`, since the address info is known statically and can be determined separately, i.e.

`Dispatcher.storage_addresses().<var_a>.address()`
instead of 
`Dispatcher.get_state_for_testing().<var_a>.address()`

And `storage_addresses` would return a trait which does not allow mutating the variables, but only returns their relative addresses.

### PR DESCRIPTION
**Pull request by Arcticae**
Closes #346

## Introduced changes

- Adds documentation for `store` and `load` cheatcodes

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: docs/src/appendix/cheatcodes/load.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# `load`
+
+> `fn load(target: ContractAddress, storage_address: felt252, size: felt252) -> Array<felt252> `
+
+Loads a `Array` of felts from `target` contract's storage, starting at `storage_address`.
```
Comment (drknzz) [inline docs/src/appendix/cheatcodes/load.md:5]:
```suggestion
Loads `size` felts from `target` contract's storage into an `Array`, starting from `storage_address`.
```

File: docs/src/appendix/cheatcodes/load.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+- `target` - Address of the contract, which storage you want to modify
+- `storage_address` - Offset of the data in the contract's storage 
+- `size` - How many felts will be loaded into the result `Array` 
+
+
+Those examples present the usage in practise:
```
Comment (drknzz) [inline docs/src/appendix/cheatcodes/load.md:12]:
```suggestion
## Example
```

File: docs/src/appendix/cheatcodes/load.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+mod Contract {
+    #[storage]
+    struct Storage {
+        plain_felt: felt252,
+        mapping: LegacyMap<felt252, felt252>,
+    }
+}
+
+// ...
+use snforge_std::{ store, load, map_entry_address };
```
Comment (drknzz) [inline docs/src/appendix/cheatcodes/load.md:24]:
```suggestion
use snforge_std::{ load, map_entry_address };
```

File: docs/src/appendix/cheatcodes/load.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+#[test]
+fn load_plain_felt() {
+    let deployed = deploy_contract();
+    let loaded = load(deployed.contract_address, selector!("plain_felt"), 1);
+    assert(loaded.len() == 1, 'Wrong loaded vector');
+    assert(*loaded.at(0) == 0, 'Wrong loaded value');
```
Comment (drknzz) [inline docs/src/appendix/cheatcodes/load.md:31]:
```suggestion
    assert(loaded.len() == 1, 'Should load 1 felt');
    assert(*loaded.at(0) == 0, 'Should load 0');
```

File: docs/src/appendix/cheatcodes/load.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+#[test]
+fn load_map_entry() {
+    let deployed = deploy_contract();
+    load(
+        deployed.contract_address, 
+        map_entry_address(
+            selector!("mapping"), // Providing variable name
+            array![123].span(),   // Providing mapping key
```
Comment (drknzz) [inline docs/src/appendix/cheatcodes/load.md:41]:
```suggestion
            array![123].span(),   // Providing mapping keys 
```
Should it be? Or another example for multiple keys, I feel like it still not clear why it's an array / how to work with nested maps

File: docs/src/appendix/cheatcodes/store.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+Stores a `Span` of felts in `target` contract's storage, starting at `storage_address`. 
+
+- `target` - Address of the contract, which storage you want to modify
+- `storage_address` - Offset of the data in the contract's storage 
+- `serialized_value` - A sequence of felts that will be inserted starting at `storage_address` 
+
+
+In order to obtain the variable address that you'd like to write to, you need to use either:
+- `selector!` macro - if the variable is a plain variable
```
Comment (drknzz) [inline docs/src/appendix/cheatcodes/store.md:13]:
```suggestion
- `selector!` macro - if the variable is a plain type
```

File: docs/src/appendix/cheatcodes/store.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+- `map_entry_address` function in tandem with `selector!` - if the variable is a mapping, and you want to insert a key-value pair
+
+Those examples present the usage in practise:
```
Comment (drknzz) [inline docs/src/appendix/cheatcodes/store.md:16]:
```suggestion
## Example
```

File: docs/src/appendix/cheatcodes/store.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    let deployed = deploy_contract();
+    store(
+        deployed.contract_address, 
+        map_entry_address(
+            selector!("mapping"), // Providing variable name
+            array![123].span(),   // Providing mapping key 
+        ),
+        array![321].span()
+    );
+}
```
Comment (drknzz) [inline docs/src/appendix/cheatcodes/store.md:47]: An assert after that, or a comment clarifying that 321 is stored at mapping[123] that would be nice

File: docs/src/appendix/cheatcodes/store.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+// ...
+use snforge_std::{ store, map_entry_address };
+
+#[test]
+fn store_plain_felt() {
+    let deployed = deploy_contract();
+    store(deployed.contract_address, selector!("plain_felt"), array![123].span());
+}
```
Comment (drknzz) [inline docs/src/appendix/cheatcodes/store.md:34]: Same here, let's have an assert / comment

File: docs/src/appendix/cheatcodes/load.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    }
+}
+
+// ...
+use snforge_std::{ store, load, map_entry_address };
+
+#[test]
+fn load_plain_felt() {
+    let deployed = deploy_contract();
+    let loaded = load(deployed.contract_address, selector!("plain_felt"), 1);
```
Comment (MaksymilianDemitraszek) [inline docs/src/appendix/cheatcodes/load.md:29]: Some example where to get the size from would be useful

File: docs/src/appendix/cheatcodes/load.md (hunk 1 above)
Comment (Arcticae) [reply docs/src/appendix/cheatcodes/load.md:5]: starting `at` is more gramatically correct
will correct though

Comment (drknzz) [reply docs/src/appendix/cheatcodes/load.md:5]: np, just some suggestions, adjust if needed

File: docs/src/appendix/cheatcodes/load.md [hunk 11]
```diff
@@ -0,0 +1,85 @@
+# `load`
+
+> `fn load(target: ContractAddress, storage_address: felt252, size: felt252) -> Array<felt252> `
+
+Loads `size` felts from `target` contract's storage into an `Array`, starting at `storage_address`.
+
+- `target` - Address of the contract, which storage you want to modify
+- `storage_address` - Offset of the data in the contract's storage 
+- `size` - How many felts will be loaded into the result `Array`
```
Comment (piotmag769) [inline docs/src/appendix/cheatcodes/load.md:9]: Should start with lowercase letter after `-`

File: docs/src/appendix/cheatcodes/load.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+#[test]
+fn load_elaborate_struct() {
+    // ...
+    load(
+        contract_address, 
+        map_entry_address(
+            selector!("mapping"), // Providing variable name
+            array![123].span(),   // Providing mapping key 
+        ),
+        StoreElaborateStructure::size().into(), // Auto-generated trait for starknet::Store deriving structs
```
Comment (piotmag769) [inline docs/src/appendix/cheatcodes/load.md:67]:
```suggestion
        StoreElaborateStructure::size().into(), // Implementation of starknet::Store for the struct generated by `#[derive(starknet::Store)]` macro
```

File: docs/src/appendix/cheatcodes/load.md [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+> 📝 **Note** 
+> 
+> The cheatcode will return 0s for uninitialized memory, which is a default storage value for Starknet contracts.
+
+
+> ⚠️ **Warning**
+> 
+> Complex data can often times be packed in a custom manner (see [this pattern](https://book.cairo-lang.org/ch99-01-03-05-optimizing-storage.html#storage-optimization-with-storepacking)) to optimize costs
+> Using different lengths for packed data might result in incorrect deserialization.
```
Comment (piotmag769) [inline docs/src/appendix/cheatcodes/load.md:83]:
```suggestion
> Complex data can often times be packed in a custom manner (see [this pattern](https://book.cairo-lang.org/ch99-01-03-05-optimizing-storage.html#storage-optimization-with-storepacking)) to optimize costs.
> If that's the case for your contract, make sure to handle deserialization properly - standard methods might not work.
```

File: docs/src/appendix/cheatcodes/load.md [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+    );
+    
+    assert(loaded.len() == 1, 'Expected 1 felt loaded');
+    assert(*loaded.at(0) == 0, 'Expected 0 value loaded');
+}
+```
+
+> 📝 **Note** 
+> 
+> The cheatcode will return 0s for uninitialized memory, which is a default storage value for Starknet contracts.
```
Comment (piotmag769) [inline docs/src/appendix/cheatcodes/load.md:77]:
```suggestion
> The cheatcode will return zeros for "uninitialized" memory, which is a default storage value for Starknet contracts.
```
https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/contract-storage/#storage_layout
If it was actually unititialized, it wouldn't be set to zeros ☝️🤓

File: docs/src/appendix/cheatcodes/load.md [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    assert(loaded.len() == 1, 'Expected 1 felt loaded');
+    assert(*loaded.at(0) == 0, 'Expected 0 value loaded');
+}
+
+#[derive(starknet::Store)]
+struct ElaborateStructure {
+    a: felt252,
+    b: felt252,
+    // ...
+}
```
Comment (piotmag769) [inline docs/src/appendix/cheatcodes/load.md:56]: How is this struct related to this example? Why is `StoreElaborateStructure` used below when we read from `LegacyMap<felt252, felt252>`?

File: docs/src/appendix/cheatcodes/store.md [hunk 16]
```diff
@@ -0,0 +1,58 @@
+# `store`
+
+> `fn store(target: ContractAddress, storage_address: felt252, serialized_value: Span<felt252>)`
+
+Stores a `Span` of felts in `target` contract's storage, starting at `storage_address`.
```
Comment (piotmag769) [inline docs/src/appendix/cheatcodes/store.md:5]: [another 🤓☝️ nit xD] it doesn't actually store a Span, it just stores felts

File: docs/src/appendix/cheatcodes/store.md [hunk 17]
```diff
@@ -0,0 +1,58 @@
+# `store`
+
+> `fn store(target: ContractAddress, storage_address: felt252, serialized_value: Span<felt252>)`
+
+Stores a `Span` of felts in `target` contract's storage, starting at `storage_address`. 
+
+- `target` - Address of the contract, which storage you want to modify
+- `storage_address` - Offset of the data in the contract's storage 
+- `serialized_value` - A sequence of felts that will be inserted starting at `storage_address`
```
Comment (piotmag769) [inline docs/src/appendix/cheatcodes/store.md:9]: Should start with lowercase after `-`

File: docs/src/appendix/cheatcodes/store.md [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+    store(
+        contract_address, 
+        map_entry_address(
+            selector!("mapping"), // Providing variable name
+            array![123].span(),   // Providing mapping key 
+        ),
+        array![321].span()
+    );
+    
+    // mapping = { 123: 321 }
```
Comment (piotmag769) [inline docs/src/appendix/cheatcodes/store.md:49]:
```suggestion
    // mapping = { 123: 321, ... }
```

File: docs/src/appendix/cheatcodes/load.md (hunk 15 above)
Comment (Arcticae) [reply docs/src/appendix/cheatcodes/load.md:56]: uhh i don't know. was rushing it obviously 😬

File: docs/src/appendix/cheatcodes/load.md (hunk 5 above)
Comment (Arcticae) [reply docs/src/appendix/cheatcodes/load.md:43]: [Add an example how to handle complex mapping key/values in the contract](https://github.com/foundry-rs/starknet-foundry/pull/1525/commits/e7b8185be349c2f96ce5c43e6ac9757672080862)

Comment (Arcticae) [reply docs/src/appendix/cheatcodes/load.md:43]: I'm not going to provide example with nested maps since it's highly unusual one, and this example range is already too extensive

File: docs/src/appendix/cheatcodes/load.md (hunk 12 above)
Comment (Arcticae) [reply docs/src/appendix/cheatcodes/load.md]: [Add explanation to how the Store trait works](https://github.com/foundry-rs/starknet-foundry/pull/1525/commits/9d5173720f4590c4bb1f74c83a6d303aa7712807)

**Review by piotmag769 — APPROVED** [review]

**Review by drknzz — APPROVED** [review]
very good documentation 👍 👍

**Comment by Arcticae** [comment]
@drknzz @piotmag769 thx guys for patience & thorough review <3
